packages/frontend/src/hooks/useKeycard.ts

```diff
@@ -1,27 +1,83 @@
-import { useLocalStorage } from "@uidotdev/usehooks";
+import { useAccount, useWalletClient } from "wagmi";
 import { generatePrivateKey, privateKeyToAccount } from "viem/accounts";
+import { getLogger } from "@logtape/logtape";
+import { useQuery, useQueryClient } from "@tanstack/react-query";
+
+import { RelayClient } from "@massmarket/client";
+import { getWindowLocation } from "@massmarket/utils";
 
 import { useShopId } from "./useShopId.ts";
 import { KeycardRole } from "../types.ts";
+import { useRelayEndpoint } from "./useRelayEndpoint.ts";
+
+const logger = getLogger(["mass-market", "frontend", "useKeycard"]);
 
-export function useKeycard(
-  keycard?: {
-    privateKey: `0x${string}`;
-    role: KeycardRole;
-    address: `0x${string}`;
-  },
-) {
-  if (!keycard) {
-    const privateKey = generatePrivateKey();
-    const account = privateKeyToAccount(privateKey);
-    keycard = {
-      privateKey,
-      role: KeycardRole.NEW_GUEST,
-      address: account.address,
-    };
-  }
+/**
+ * This hook returns a keycard for the browser session.
+ * The async queryFn where we enroll the keycard only executes once unless variables in the queryKey changes.
+ * The keycard will be cached for the duration of the browser session regardless of refreshes.
+ */
+export function useKeycard(role: KeycardRole = "guest") {
   const { shopId } = useShopId();
-  const keyCardID = "keycard" + shopId;
-  // console.log({shopId, keyCardID});
-  return useLocalStorage(keyCardID, keycard);
+  const { data: wallet } = useWalletClient();
+  const { address } = useAccount();
+  const { relayEndpoint } = useRelayEndpoint();
+  const queryClient = useQueryClient();
+
+  return useQuery({
+    // queryFn will not execute till these variables are defined.
+    enabled: !!shopId && !!wallet && !!relayEndpoint && !!address,
+    queryKey: [
+      "keycard",
+      address,
+      role,
+    ],
```

Comment:
oh wait! we need what happen to `shopid` ? we defaintly need that to be in the key. A merchant keycard on one shop will not work on a different shop